Update & Expand Meta Tags

solonovamax opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

## Summary

Improves/expands the meta tags for shared files.

- Adds the following opengraph tags, depending on the file type
  - images:
    - `og:image:type`: the mimetype of the image file (**TODO**: should this be added for all shared files? all files will always have an `og:image` tag with either the preview or the nc logo)
  - audio:
    - `og:audio`: a direct link to the audio file
    - `og:audio:type`: the mimetype of the audio file
  - video:
    - `og:video`: a direct link to the video file
    - `og:video:type`: the mimetype of the video file
- Updates the `og:type` tag from `object` (which is not a type that exists, as far as I can tell) to `website`
- Adds meta tags for twitter
  - `twitter:title`: same as `og:title`
  - `twitter:description`: same as `og:description`
  - `twitter:card`: `summary_large_image` if the shared file is an image & it has a preview, otherwise `summary`
  - `twitter:image`: same as `og:image`

### Before changes

![Image](https://github.com/user-attachments/assets/ca761efc-9966-4481-a479-d6cc7b7f29c7)


### After changes

> [!NOTE]
> I tested this on my personal self-hosted nextcloud, which is running nextcloud `29.0.10`. I did this by just manually editing the `apps/files_sharing/lib/DefaultPublicShareTemplateProvider.php` file.

![image](https://github.com/user-attachments/assets/42dc8993-0594-40f6-9f33-b196521d33a0)


## Checklist

- [x] Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [x] [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

solonovamax committed:
feat(files_sharing): Add twitter meta tags

Adds the following twitter meta tags
- `twitter:card`: `summary_large_image` if the shared file is an image & it has a preview, otherwise `summary`
- `twitter:title`: same as `og:title`
- `twitter:description`: same as `og:description`
- `twitter:image`: same as `og:image`

Fixes nextcloud/server#49871

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solracsf commented:
Can you please fix your commit?
https://www.conventionalcommits.org/en/v1.0.0/#summary

solracsf commented on apps/files_sharing/lib/DefaultPublicShareTemplateProvider.php:
```diff
@@ -226,11 +227,24 @@ protected function addOpenGraphHeaders(IShare $share): void {
+		// Twitter Support: https://developer.x.com/en/docs/x-for-websites/cards/overview/markup
+		Util::addHeader('meta', ['property' => 'twitter:title', 'content' => $shareNode->getName()]);
+		Util::addHeader('meta', ['property' => 'twitter:description', 'content' => $description]);
+		if ($hasImagePreview) {
+			Util::addHeader('meta', ['property' => 'twitter:card', 'content' => 'summary_large_image']);
+		} else {
+			Util::addHeader('meta', ['property' => 'twitter:card', 'content' => 'summary']);
+		}
```
```suggestion
		Util::addHeader('meta', ['property' => 'twitter:card', 'content' => $hasImagePreview ? 'summary_large_image' : 'summary']);
```

solonovamax replied:
ah, wasn't sure if php had ternary expressions (not a php dev lol)

I can push a fix in a few.

solracsf commented on apps/files_sharing/lib/DefaultPublicShareTemplateProvider.php:
```diff
@@ -226,11 +227,24 @@ protected function addOpenGraphHeaders(IShare $share): void {
 		}
 
+		$description = $this->defaults->getName() . ($this->defaults->getSlogan() !== '' ? ' - ' . $this->defaults->getSlogan() : '');
+
```
```suggestion
$title = $shareNode->getName();
```
And reuse `$title` for bot OG and Twitter title headers.

solonovamax replied:
I chose not to reuse it, as it's just a simple method call, whereas the description has some logic associated with it that I didn't want to duplicate.

but I can pull it out into a variable if you'd like.

solonovamax commented:
Applied requested changes.

solracsf commented on apps/files_sharing/lib/DefaultPublicShareTemplateProvider.php:
```diff
@@ -190,15 +189,17 @@ public function renderPage(IShare $share, string $token, string $path): Template
 		if ($this->previewManager->isMimeSupported($shareNode->getMimetype())) {
 			// For images we can use direct links
 			if ($shareNode->getMimePart() === 'image') {
+			    $hasImagePreview = true;
```
```suggestion
				$hasImagePreview = true;
```

solonovamax committed:
feat(files_sharing): Add opengraph meta tags for multimedia, change opengraph type to website

- Adds the following opengraph tags
  - images:
    - `og:image:type`: the mimetype of the image file
  - audio:
    - `og:audio`: a direct link to the audio file
    - `og:audio:type`: the mimetype of the audio file
  - video:
    - `og:video`: a direct link to the video file
    - `og:video:type`: the mimetype of the video file
- Changes th `og:type` meta tag from `object` (which is not valid) to `website`

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

solonovamax commented:
I've also added some additional changes to the opengraph meta tags. I will update the title/description in 15-20 with the additional changes.

solonovamax commented:
oops, accidentally left in some funny formatting. I'll fix that in a bit.

other than that, the other failing checks seem to be either flaky tests, broken tests, or checks which won't run on forks.

come-nc commented on apps/files_sharing/lib/DefaultPublicShareTemplateProvider.php:
```diff
@@ -226,11 +227,34 @@ protected function addOpenGraphHeaders(IShare $share): void {
+
+
+		// Twitter Support: https://developer.x.com/en/docs/x-for-websites/cards/overview/markup
+		Util::addHeader('meta', ['property' => 'twitter:title', 'content' => $title]);
+		Util::addHeader('meta', ['property' => 'twitter:description', 'content' => $description]);
+		Util::addHeader('meta', ['property' => 'twitter:card', 'content' => $hasImagePreview ? 'summary_large_image' : 'summary']);
+		Util::addHeader('meta', ['property' => 'twitter:image', 'content' => $ogPreview]);
```
Why are we adding twitter specific tags? Can’t it read the standard tags?

solracsf replied:
They're not optimized. See screenshots in 1st post.

solonovamax replied:
> Why are we adding twitter specific tags? Can’t it read the standard tags?

It can, however imo it doesn't look as nice. yes, technically only the `twitter:card` tag is *needed* as it will fall back to the opengraph tags, however there is no *harm* in including them. (& there may be some services which only support the twitter tags, or if one of the twitter tags is present, it only uses twitter tags. unsure, as I have not tested this)

solonovamax commented:
Any updates on the review?

sorbaugh commented:
Re-requesting reviews..

solonovamax commented:
this was merged into 31.0.0, correct? bc on my nextcloud instance which is on version 31.0.7, I'm not seeing these changes

susnux commented:
No its merged for Nextcloud 32 as you can see by the milestone
![Screenshot_20250816_123245_Firefox](https://github.com/user-attachments/assets/a358e445-0d6d-478c-82ea-8cebcf46c78c)

solonovamax commented:
> No its merged for Nextcloud 32 as you can see by the milestone
> ![Screenshot_20250816_123245_Firefox](https://github.com/user-attachments/assets/a358e445-0d6d-478c-82ea-8cebcf46c78c)

ah, it doesn't show that to me on mobile, which is probably why. I could only see where it was mentioned, and it was only mentioned in PRs for nc 31

well, hopefully nc 32 will happen sometime soon, bc it'll make sharing nc links to other social media much nicer

susnux commented:
> well, hopefully nc 32 will happen sometime soon, bc it'll make sharing nc links to other social media much nicer

Yes: 2025-09-27
See also https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule#nextcloud-32
